flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/AsyncTableFunction.java

```diff
@@ -0,0 +1,139 @@
+ * operation, its context is stored in the operator immediately after invoking "eval",
+ * avoiding blocking for each stream input as long as the internal buffer is not full.
+ *
+ * <p>{@link CompletableFuture} can be passed into callbacks or futures to collect the result data.
+ * An error can also be propagate to the async IO operator by
+ * {@link CompletableFuture#completeExceptionally(Throwable)}.
+ *
+ * <p>User-defined functions must have a default constructor and must be instantiable during
+ * runtime.
+ *
+ * <p>By default the result type of an evaluation method is determined by Flink's type extraction
+ * facilities. This is sufficient for basic types or simple POJOs but might be wrong for more
+ * complex, custom, or composite types. In these cases {@link TypeInformation} of the result type
+ * can be manually defined by overriding {@link #getResultType}.
+ *
+ * <p>Internally, the Table/SQL API code generation works with primitive values as much as possible.
+ * If a user-defined table function should not introduce much overhead during runtime, it is
+ * recommended to declare parameters and result types as primitive types instead of their boxed
+ * classes. DATE/TIME is equal to int, TIMESTAMP is equal to long.
+ *
+ * <p>Example:
+ *
+ * {@code
+ *
+ *   public class HBaseAsyncTableFunction extends AsyncTableFunction<String> {
+ *
+ *     // implement an "eval" method with as many parameters as you want
+ *     public void eval(CompletableFuture<Collection<String>> result, String rowkey) {
+ *       Get get = new Get(Bytes.toBytes(rowkey));
+ *       ListenableFuture<Result> future = hbase.asyncGet(get);
+ *       Futures.addCallback(future, new FutureCallback<Result>() {
+ *         public void onSuccess(Result result) {
+ *           List<String> ret = process(result);
+ *           result.complete(ret);
+ *         }
+ *         public void onFailure(Throwable thrown) {
+ *           result.completeExceptionally(thrown);
+ *         }
+ *       });
+ *     }
+ *
+ *     // you can overload the eval method here ...
+ *   }
+ * }
+ *
+ * <p>NOTE: the {@link AsyncTableFunction} can not be used as UDTF currently. It only used in
+ * temporal table join as an async lookup function.
+ *
+ * @param <T> The type of the output row
+ */
+@Experimental
+public abstract class AsyncTableFunction<T> extends UserDefinedFunction {
+
+	/**
+	 * Returns the result type of the evaluation method with a given signature.
+	 *
+	 * <p>This method needs to be overridden in case Flink's type extraction facilities are not
+	 * sufficient to extract the {@link TypeInformation} based on the return type of the evaluation
+	 * method. Flink's type extraction facilities can handle basic types or
+	 * simple POJOs but might be wrong for more complex, custom, or composite types.
+	 *
+	 * @return {@link TypeInformation} of result type or <code>null</code> if Flink should determine the type
+	 */
+	public TypeInformation<T> getResultType() {
```

Comment:
Use `DataType` instead?

Reply:
Currently, we do not rely on this method `getResultType` and  `getParameterTypes`, we can get the result type from `TableSource#getResultType` or `TableSource#getProducedDataType`. 

So I removed these methods, and will introduce new methods using `DataType` after `TableFunction`  is refactored.

Reply:
I don't think so. `AsyncTableFunction` should first be a table function, it can be treated as normal table function but executed in an aync way. You current solution will make this interface can only apply to lookup join which i think is somehow coupled. 
We still need to treat these two things separately.

Reply:
Do you mean users have to implement  `AsyncTableFunction#getResultType` even if they have provide the information in `TableSource#getResultType`?

The only concern I have is that the `DataType getResultType()` is not a final solution. According to ["Type System Part II"](https://docs.google.com/document/d/1RM_-XvD25AldUOl7xRSvA3sascWBa1-Jze3i5mvb7WU/edit#), we need to introduce `getOutputTypeInference(): OutputTypeInference` to satisfy all the requirements. Which means even though we expose `DataType getResultType()`, it will also be deprecated and refactored in the near future.

Reply:
We can at least keep align with current interfaces until the type system part 2 begins.

Reply:
So let's keep `TypeInformation<T> getResultType()` ?